src/coreclr/jit/gentree.cpp

```diff
@@ -2620,29 +2620,55 @@ unsigned Compiler::gtSetMultiOpOrder(GenTreeMultiOp* multiOp)
     unsigned level  = 0;
     unsigned lvl2   = 0;
 
-#if defined(FEATURE_HW_INTRINSICS) && defined(TARGET_XARCH)
-    if (multiOp->OperIs(GT_HWINTRINSIC) && (multiOp->GetOperandCount() == 1) &&
-        multiOp->AsHWIntrinsic()->OperIsMemoryLoadOrStore())
+#if defined(FEATURE_HW_INTRINSICS)
+    if (multiOp->OperIs(GT_HWINTRINSIC))
     {
-        costEx = IND_COST_EX;
-        costSz = 2;
+        GenTreeHWIntrinsic* hwTree = multiOp->AsHWIntrinsic();
+#if defined(TARGET_XARCH)
+        if ((hwTree->GetOperandCount() == 1) && hwTree->OperIsMemoryLoadOrStore())
+        {
+            costEx = IND_COST_EX;
+            costSz = 2;
 
-        GenTree* addr = multiOp->Op(1)->gtEffectiveVal();
-        level         = gtSetEvalOrder(addr);
+            GenTree* addr = hwTree->Op(1)->gtEffectiveVal();
+            level         = gtSetEvalOrder(addr);
 
-        // See if we can form a complex addressing mode.
-        if (addr->OperIs(GT_ADD) && gtMarkAddrMode(addr, &costEx, &costSz, multiOp->TypeGet()))
-        {
-            // Nothing to do, costs have been set.
+            // See if we can form a complex addressing mode.
+            if (addr->OperIs(GT_ADD) && gtMarkAddrMode(addr, &costEx, &costSz, hwTree->TypeGet()))
+            {
+                // Nothing to do, costs have been set.
+            }
+            else
+            {
+                costEx += addr->GetCostEx();
+                costSz += addr->GetCostSz();
+            }
+
+            hwTree->SetCosts(costEx, costSz);
+            return level;
         }
-        else
+#endif
+        switch (hwTree->gtHWIntrinsicId)
         {
-            costEx += addr->GetCostEx();
-            costSz += addr->GetCostSz();
+#if defined(TARGET_XARCH)
+            case NI_Vector128_Create:
+            case NI_Vector256_Create:
+#elif defined(TARGET_ARM64)
+            case NI_Vector64_Create:
+            case NI_Vector128_Create:
+#endif
+            {
+                if (hwTree->gtGetOp1()->OperIsConst() && (hwTree->gtGetOp2() == nullptr))
+                {
+                    // Vector.Create(cns) is cheap but not that cheap to be (1,1)
+                    costEx = 2;
```

Comment:
This should be `IND_COST_EX`

Reply:
Actually, it probably needs to be a "bit" more complex.

If there is one operand and all operands are constant and its not representing `all bits zero` or `all bits set` then its `IND_COST_EX`.

If part of the value isn't `constant` then the cost increases as the number of operands increases. We don't currently, but could eventually, handle "partial constants".

If the value represents `all bits zero` or `all bits set`, then its cheaper and its just `xor` or the relevant `cmp` SIMD instruction and is special cased by hardware.

Reply:
@tannergooding yes I had "all zeros/ones" cases in mind but the problem that they complicate code a lot (especially the AllBitsSet case for different types) for a very rare case where usually get_Zero/get_AllBitsSet intrinsics are used. I think it won't hurt if we do CSE more often for these cases or we better move the logic to recognize  get_Zero/get_AllBitsSet early in morph/importer and it will work as expected + IR will be simplified earlier.